sdk/src/Core/Amazon.Runtime/Internal/Util/StreamingUtf8JsonReader.cs

```diff
@@ -0,0 +1,169 @@
+                if (_reader.IsFinalBlock)
+                {
+                    ArrayPool<byte>.Shared.Return(_buffer);
+                    break;
+                }
+
+                GetMoreBytesFromStream(_stream, ref _buffer, ref _reader);
+
+                hasMoreData = _reader.Read();
+            }
+
+            return hasMoreData;
+        }
+
+        private static void GetMoreBytesFromStream(Stream stream, ref byte[] buffer, ref Utf8JsonReader reader)
+        {
+            int bytesRead = 0;
+            ReadOnlySpan<byte> leftover = buffer.AsSpan().Slice((int)reader.BytesConsumed);
+            if (reader.BytesConsumed < buffer.Length)
+            {
+                // If BytesConsumed is 0 that means that the previous Read failed because the JSON token was too large to fit in the buffer.
+                // In that case we need to resize the buffer and try again to read the JSON token.
+                if (reader.BytesConsumed == 0)
+                {
+                    var resizedBuffer = ArrayPool<byte>.Shared.Rent(Math.Min(int.MaxValue, (buffer.Length * 2)));
+                    Logger.GetLogger(typeof(StreamingUtf8JsonReader)).DebugFormat("Resizing buffer from {0} to {1}", buffer.Length, resizedBuffer.Length);
+                    // copy over the data from the previous read's buffer to the newly resized buffer.
+                    buffer.AsSpan().CopyTo(resizedBuffer);
+                    // return the previous buffer to the pool and set the new buffer to equal the resized buffer.
+                    ArrayPool<byte>.Shared.Return(buffer);
+                    buffer = resizedBuffer;
+                    // fill the new resized buffer with data from the stream. the offset MUST be leftover.Length 
+                    // so we don't overwrite the data that was copied from the previous buffer, and the number of bytes
+                    // we read must be buffer.Length - leftover.Length which is just the second half of the buffer.
+                    bytesRead = FillBuffer(stream, ref buffer, leftover.Length, buffer.Length - leftover.Length);
+                    var resizedSpan = buffer.AsSpan(0, bytesRead + leftover.Length);
+                    reader = new Utf8JsonReader(resizedSpan, isFinalBlock: bytesRead == 0, reader.CurrentState);
+                    // early return since we have a reader
+                    return;
+                }
+                else
+                {
+                    // Move the unprocessed data from the buffer to the start and fill
+                    // remaining space in the buffer with new content from the stream.
+                    leftover.CopyTo(buffer);
+                    bytesRead = FillBuffer(stream, ref buffer, leftover.Length, buffer.Length - leftover.Length);
+                }
+            }
+            else
+            {
+                bytesRead = FillBuffer(stream, ref buffer, 0, buffer.Length);
+            }
+
+            if (bytesRead == 0)
```

Comment:
I'm thinking this is getting complicated because we are depending on if the code goes into the `else` block up above dealing with leftover data we also expect this code to handle how to create the reader.  What do you think of this version where whenever we go into any condition block it does everything it needs to do and not rely on logic somewhere else to finish things up.

I also updated how the `Utf8JsonReader` was created in the resize condition to say if it is the final block if we did not fill up the entire new buffer.

```csharp
private static void GetMoreBytesFromStream(Stream stream, ref byte[] buffer, ref Utf8JsonReader reader)
{
    if (reader.BytesConsumed < buffer.Length)
    {
        ReadOnlySpan<byte> leftover = buffer.AsSpan().Slice((int)reader.BytesConsumed);

        // If BytesConsumed is 0 that means that the previous Read failed because the JSON token was too large to fit in the buffer.
        // In that case we need to resize the buffer and try again to read the JSON token.
        if (reader.BytesConsumed == 0)
        {
            var resizedBuffer = ArrayPool<byte>.Shared.Rent(Math.Min(int.MaxValue, (buffer.Length * 2)));
            Logger.GetLogger(typeof(StreamingUtf8JsonReader)).DebugFormat("Resizing buffer from {0} to {1}", buffer.Length, resizedBuffer.Length);
            // copy over the data from the previous read's buffer to the newly resized buffer.
            buffer.AsSpan().CopyTo(resizedBuffer);
            // return the previous buffer to the pool and set the new buffer to equal the resized buffer.
            ArrayPool<byte>.Shared.Return(buffer);
            buffer = resizedBuffer;
            // fill the new resized buffer with data from the stream. the offset MUST be leftover.Length 
            // so we don't overwrite the data that was copied from the previous buffer, and the number of bytes
            // we read must be buffer.Length - leftover.Length which is just the second half of the buffer.
            var bytesRead = FillBuffer(stream, ref buffer, leftover.Length, buffer.Length - leftover.Length);
            var resizedSpan = buffer.AsSpan(0, bytesRead + leftover.Length);
            reader = new Utf8JsonReader(resizedSpan, isFinalBlock: bytesRead == 0 || resizedSpan.Length != resizedBuffer.Length, reader.CurrentState);
            // early return since we have a reader
            return;
        }
        // The buffer has had some data processed but there is leftover data unprocessed.
        // In this case we move the unprocessed data to the start of the buffer and fill the rest with new data.
        else
        {
            leftover.CopyTo(buffer);
            var bytesRead = FillBuffer(stream, ref buffer, leftover.Length, buffer.Length - leftover.Length);
            reader = new Utf8JsonReader(buffer.AsSpan(0, bytesRead + leftover.Length), isFinalBlock: ((bytesRead + leftover.Length) != buffer.Length || bytesRead == 0), reader.CurrentState);
            return;
        }
    }
    // In this case enture buffer has been processed so attempt to refill the enture
    // buffer.
    else
    {
        var bytesRead = FillBuffer(stream, ref buffer, 0, buffer.Length);
        reader = new Utf8JsonReader(buffer.AsSpan(0, bytesRead), isFinalBlock: bytesRead == 0, reader.CurrentState);
        return;
    }
}
```